### DISCUSSION TITLE
Using prebuilt SD actions with custom actions

### DISCUSSION THREAD

**Discussion by ForgetfulMemoryFoam** [elgatosf/elgato#115] [Help] [answered]
Hello all!

Is there a way to integrate an already pre-built SD action, such as a hotkey, with some custom code that I build? I want to build something simple for a customer that doesn't involve them needing to set up any hotkeys, but they will be created programmatically. That way they can delete the actions and not worry about re-importing the profile if they change their mind.

**Comment by ChekTek** [comment]
Hi @Alex-M-Howard 

I don't believe what you are asking is directly possible, without recreating a "hotkey" action yourself. If you do recreate the hotkey action yourself, you could then bundle a profile in the plugin that includes these hotkeys preconfigured. You could then keep track of the actions using the `willAppear` and `willDisappear` events and dynamically update them as you see fit. 

We are working on something we're calling "Full Canvas Control" that may make this easier in the future. It will essentially allow one action to take over all of the Stream Deck canvas, which would make it easier for you to dynamically populate the hotkeys.

**Reply by ForgetfulMemoryFoam** [reply]
Thanks for getting back to me @ChekTek. 

I might make a profile that is already set up, but I was hoping for a way to create a plugin so that if someone deleted it, they wouldn't have to reimport the profile again. The target audience I am going for will include people that aren't tech savvy and I want as little work on their part. My current workaround, I created a separate binary that contains a rust API server that will press the key combinations I want.

I'm definitely looking forward to the full canvas control. The other plugin I've made here at work will greatly benefit from this, as there are tons of actions and possibilities that could make it easier to set up.

**Reply by ChekTek** [reply, marked as answer]
If you bundle a profile into your plugin, when you call `switchToProfile` it will either switch to the profile if it's installed or prompt the user to install the profile if it's not already there. Creating an action that does this may be your best bet. It will still require an extra click from the user, but then even if the profile is deleted it will just prompt them to reinstall when it is needed. 

[switchToProfile](https://docs.elgato.com/sdk/plugins/events-sent#switchtoprofile)

If you want to see this in action, you can check out the [Volume Controller](https://marketplace.elgato.com/product/volume-controller-34d9aa59-a41a-4a4c-a853-202ca91409e1) plugin's "Auto Detection" action. This uses a profile built into a plugin (we will be converting it over to the Full Canvas Control once the API is available). 

We kind of recommend trying to include all dependencies inside of the plugin itself, as Stream Deck can control the entire life cycle (i.e. restarting the plugin if it crashes etc). Speaking from experience (we currently have a separate audio server running for Volume Controller), it can be a real pain having a separate process providing this functionality. I'm not sure if you have started using the new [Node.js SDK](https://www.npmjs.com/package/@elgato/cli?activeTab=readme) yet, but I would highly recommend looking into [node-gyp](https://github.com/nodejs/node-gyp) addons. Typically, these Node.js addons are built with C++ and to be honest they are a bit tricky to set up, but I believe there is even a way to build them in Rust! I haven't tried it yet, but I'm also looking forward to giving Node.js [Rust addons](https://github.com/napi-rs/napi-rs) a try.

**Reply by ForgetfulMemoryFoam** [reply]
I didn't realize you could bundle profiles directly into the plugin. I'll definitely check this out. Thank you!
